### ISSUE TITLE
[Feature Request] Add warnings if train data is not normalized/standardized

### ISSUE THREAD

**Issue by Balandat** [pytorch/botorch#208]
# 🚀 Feature Request

Let's emit a warning if input data to the standard models are not normalized/standardized.

## Motivation

Our standard models assume that the input data is normalized (train inputs in the unit cube, train targets with zero mean and unit variance). 
While we do indicate this in the tutorials/docstrings, a good amount of the issues we see reported is b/c users do not standardize the data.

**Comment by Balandat** [comment]
Let's also check for NaNs in the inputs - right now doing this will result in very confusing errors deep down in the stack.

**Comment by mc-robinson** [comment]
Hi @Balandat, I was prepping to start submitting a PR related to #180, and realized that normalization/standardization would be really important -- so decided to start and try to help with this issue. I've written some draft functions shown below, but I can't say I'm confident where to place them without fully understanding the batching of the inputs. My initial guess is that the checks should be run somewhere in the `__init__` functions of the gp_regression models; however, I'm sure you know much better. 

I've written three quick functions for checking the input:

```python
import warnings

def check_X_no_nans(X: Tensor) -> bool:
    return not torch.any(torch.isnan(X)).item() > 0

def check_X_normalization(X: Tensor, atol: float = 1e-2) -> bool:
    # is X checked to have two dims anywhere
    if (not torch.allclose(torch.min(X, dim=0)[0], torch.zeros(X.shape[-1]), atol=atol) or
        not torch.allclose(torch.max(X, dim=0)[0], torch.ones(X.shape[-1]), atol=atol)):
        warnings.warn("The input `train_X` has not been normalized to the unit cube. "
                      "Please consider min-max scaling the input data.",
                      stacklevel=2)
        return False
    else: 
        return True

def check_y_standardization(y: Tensor, atol: float = 1e-2) -> bool:
    if (not torch.allclose(torch.mean(y, dim=0), torch.zeros(y.shape[-1]), atol=atol) or
        not torch.allclose(torch.std(y, dim=0), torch.ones(y.shape[-1]), atol=atol)):
        warnings.warn("The target `train_y` has not been standardized. "
                      "Please consider scaling the target to mean 0 and standard deviation 1.",
                      stacklevel=2)
        return False
    else:
        return True
```

One thing I couldn't find is if `train_X` is checked to have at least two dimensions anywhere? I'm not sure if an error is ever thrown if X has only one dimension (I know broadcasting is sometimes a huge pain with PyTorch loss functions).

Anyway, let me know if any of those functions look useful and where they should be placed if so (perhaps in `botorch/botorch/models/`?)

I also went ahead and wrote tests to make sure the warnings work correctly, that might be useful if this gets turned into a PR

```python
import unittest

class TestInputWarnings(unittest.TestCase):
  
    def test_nan_error(self):
        i = torch.Tensor([[8,7,6,9.34,20],
                          [3,4,float(np.nan),5,4]])
        self.assertEqual(check_X_no_nans(i), False)
        
        i2 = torch.Tensor([[8,7,6,9.34,20],
                          [3,4,0.0,5,4]]) 
        self.assertEqual(check_X_no_nans(i2), True)
        
    def test_normalization_warning(self):
      
        # check warning is thrown
        i = torch.randn((10, 3))
        with warnings.catch_warnings(record=True) as w:
            # Ensure warnings are being shown
            warnings.simplefilter("always")
            # Trigger Warning
            check_X_normalization(i)
            # Check warning occurs
            self.assertEqual(len(w), 1)
            self.assertIn('Please consider min-max scaling the input data.', str(w[0]))
            
        # check warning is not thrown after normalization
        i_norm = (i - i.min(dim=0)[0])/(i.max(dim=0)[0] - i.min(dim=0)[0])
        with warnings.catch_warnings(record=True) as w:
            # Ensure warnings are being shown
            warnings.simplefilter("always")
            # Try to trigger Warning
            check_X_normalization(i_norm)
            # Check no warning occurs
            self.assertEqual(len(w), 0)
            
    def test_standardization_warning(self):
      
        # check warning is thrown 
        t = torch.rand((100,))
        with warnings.catch_warnings(record=True) as w:
            # Ensure warnings are being shown
            warnings.simplefilter("always")
            # Trigger Warning
            check_y_standardization(t)
            # Check warning occurs
            self.assertEqual(len(w), 1)
            self.assertIn('The target `train_y` has not been standardized.', str(w[0]))
            
        # check warning is not thrown after standardization
        t_std = (t-t.mean(dim=0))/(t.std(dim=0))
        with warnings.catch_warnings(record=True) as w:
            # Ensure warnings are being shown
            warnings.simplefilter("always")
            # Try to trigger Warning
            check_y_standardization(t_std)
            # Check no warning occurs
            self.assertEqual(len(w), 0) 
```

**Comment by Balandat** [comment]
Thanks @mc-robinson for drafting these helper functions and the tests. Yes, these should indeed go into the constructors of the models, I'll find the right spot for them.

One thing we'll want to do is allow the option for not checking whether the data standardized (people may have a legitimate reason to do so, or it makes thinks less awkward when writing unit tests..). We should probably introduce a new setting in `botorch/settings.py` for this, as this should be more generally useful in other places.

Also, we'll want to make the tests less brittle w.r.t. the warning assertions (see discussion on https://github.com/cornellius-gp/gpytorch/issues/824), but I can take care of that.

**Comment by Balandat** [comment]
Oh, and the natural place for the helper functions to live would be `botorch/models/utils.py`

**Comment by mc-robinson** [comment]
Ok great thanks @Balandat, just let me know if there is anything I can do to help. And good to know about the warning assertions! A bit ashamed to say I had PyTorch PR merged using the length of the warnings in the tests (copying their style), perhaps somebody needs to go in and fix all of those!

**Comment by Balandat** [comment]
Well, if it's sufficiently high up in the stack and as long as PyTorch tightly controls all the warnings, that's something not overly unreasonable. More of a concern to us who have to work with potentially different versions of the deps.

**Comment by wbkdef** [comment]
I'm not getting a warning and it seems to be working fine for un-normalized data.  Has this been removed?  Normalization is no longer needed?

**Comment by Balandat** [comment]
@wbkdef what model are you using? The standard ones should emit a warning if the data is not normalized/standardized. Maybe there is some warning suppression going on in your code?
